File: firebase/functions/lib/on_call_function.dart
import 'dart:async';

import 'package:firebase_functions_interop/firebase_functions_interop.dart'
    hide CloudFunction;
import 'cloud_function.dart';
import 'utils/firestore_utils.dart';
import 'package:data_models/cloud_functions/requests.dart';

/// Class for onCall methods. These are requests that are mostly called from
/// our client apps.
///
/// Request types should be defined in the shared cloudfunctions folder.
abstract class OnCallMethod<T extends SerializeableRequest>
    implements CloudFunction {
  @override
  final String functionName;

  T Function(dynamic data) requestFromData;

  final RuntimeOptions? runWithOptions;

  OnCallMethod(
    this.functionName,
    this.requestFromData, {
    this.runWithOptions,
  });

  Future<dynamic> action(T request, CallableContext context);

  Future<dynamic> callAction(dynamic data, CallableContext context) async {
    try {
      print('getting request');
      final request = requestFromData(firestoreUtils.fromFirestoreJson(data));

      // Print the request for debug purposes
      print(request);

      final result = await action(request, context);
      return result ?? {};
    } catch (e, stacktrace) {
      print('Error during action $functionName');
      print(e);
      print(stacktrace);
      rethrow;
    }
  }

  @override
  void register(FirebaseFunctions functions) {
    functions[functionName] = functions
        .runWith(
          runWithOptions ??
              RuntimeOptions(
                timeoutSeconds: 60,
                memory: '1GB',
                minInstances: 0,
              ),
        )
        .https
        .onCall(callAction);
  }
}

File: firebase/functions/src/on_call_function.ts
import * as functions from 'firebase-functions';
import { firestoreUtils } from './utils/infra/firestore_utils';

export interface RuntimeOptions {
  timeoutSeconds?: number;
  memory?: string;
  minInstances?: number;
}

/**
 * Abstract base class for onCall Cloud Functions.
 * Corresponds to Dart's OnCallMethod<T>.
 *
 * Usage: extend this class and implement action().
 */
export abstract class OnCallMethod<T> {
  readonly functionName: string;
  readonly requestFromData: (data: unknown) => T;
  readonly runWithOptions?: RuntimeOptions;

  constructor(
    functionName: string,
    requestFromData: (data: unknown) => T,
    opts?: { runWithOptions?: RuntimeOptions }
  ) {
    this.functionName = functionName;
    this.requestFromData = requestFromData;
    this.runWithOptions = opts?.runWithOptions;
  }

  abstract action(request: T, context: functions.https.CallableContext): Promise<unknown>;

  async callAction(data: unknown, context: functions.https.CallableContext): Promise<unknown> {
    try {
      console.log('getting request');
      const request = this.requestFromData(firestoreUtils.fromFirestoreJson(data as Record<string, unknown>));
      console.log(request);
      const result = await this.action(request, context);
      return result ?? {};
    } catch (e) {
      console.error(`Error during action ${this.functionName}`, e);
      throw e;
    }
  }

  register(): functions.HttpsFunction {
    const opts = this.runWithOptions ?? { timeoutSeconds: 60, memory: '1GB', minInstances: 0 };
    return functions
      .runWith(opts as functions.RuntimeOptions)
      .https.onCall((data, context) => this.callAction(data, context));
  }
}
